diff -uN -r orig/src/libgcrypt.c mod/src/libgcrypt.c
--- a/src/libgcrypt.c	2024-10-16 08:03:11.000000000 +0000
+++ b/src/libgcrypt.c	2026-08-29 21:54:57.184629326 +0000
@@ -218,6 +218,121 @@
     return (rc == 0) ? 0 : -1;
 }
 #endif
+
+#if LIBSSH2_RSA_SHA2
+int
+_libssh2_rsa_sha2_verify(libssh2_rsa_ctx * rsa,
+                         size_t hash_len,
+                         const unsigned char *sig,
+                         size_t sig_len,
+                         const unsigned char *m, size_t m_len)
+{
+    unsigned char hash[SHA512_DIGEST_LENGTH];
+    const char *algo = NULL;
+    gcry_sexp_t s_sig = NULL, s_hash = NULL;
+    int rc = -1;
+
+    if(hash_len == SHA256_DIGEST_LENGTH) {
+        gcry_md_hash_buffer(GCRY_MD_SHA256, hash, m, m_len);
+        algo = "sha256";
+    }
+    else if(hash_len == SHA512_DIGEST_LENGTH) {
+        gcry_md_hash_buffer(GCRY_MD_SHA512, hash, m, m_len);
+        algo = "sha512";
+    }
+    else {
+        return -1;
+    }
+
+    if(gcry_sexp_build(&s_hash, NULL,
+                       "(data (flags pkcs1) (hash %s %b))",
+                       algo, hash_len, hash)) {
+        goto error;
+    }
+
+    if(gcry_sexp_build(&s_sig, NULL, "(sig-val(rsa(s %b)))", sig_len, sig)) {
+        goto error;
+    }
+
+    rc = gcry_pk_verify(s_sig, s_hash, rsa);
+
+error:
+    if(s_sig)
+        gcry_sexp_release(s_sig);
+    if(s_hash)
+        gcry_sexp_release(s_hash);
+
+    return (rc == 0) ? 0 : -1;
+}
+
+int
+_libssh2_rsa_sha2_sign(LIBSSH2_SESSION * session,
+                       libssh2_rsa_ctx * rsactx,
+                       const unsigned char *hash,
+                       size_t hash_len,
+                       unsigned char **signature, size_t *signature_len)
+{
+    gcry_sexp_t sig_sexp = NULL;
+    gcry_sexp_t data = NULL;
+    const char *algo = NULL;
+    const char *tmp;
+    int rc;
+    size_t size;
+
+    if(hash_len == SHA256_DIGEST_LENGTH)
+        algo = "sha256";
+    else if(hash_len == SHA512_DIGEST_LENGTH)
+        algo = "sha512";
+    else
+        return -1;
+
+    if(gcry_sexp_build(&data, NULL,
+                       "(data (flags pkcs1) (hash %s %b))",
+                       algo, hash_len, hash)) {
+        return -1;
+    }
+
+    rc = gcry_pk_sign(&sig_sexp, data, rsactx);
+
+    gcry_sexp_release(data);
+
+    if(rc) {
+        return -1;
+    }
+
+    data = gcry_sexp_find_token(sig_sexp, "s", 0);
+    if(!data) {
+        gcry_sexp_release(sig_sexp);
+        return -1;
+    }
+
+    tmp = gcry_sexp_nth_data(data, 1, &size);
+    if(!tmp) {
+        gcry_sexp_release(data);
+        gcry_sexp_release(sig_sexp);
+        return -1;
+    }
+
+    if(tmp[0] == '\0') {
+        tmp++;
+        size--;
+    }
+
+    *signature = LIBSSH2_ALLOC(session, size);
+    if(!*signature) {
+        gcry_sexp_release(data);
+        gcry_sexp_release(sig_sexp);
+        return -1;
+    }
+    memcpy(*signature, tmp, size);
+    *signature_len = size;
+
+    gcry_sexp_release(data);
+    gcry_sexp_release(sig_sexp);
+
+    return 0;
+}
+#endif
 #endif
 
 #if LIBSSH2_DSA
diff -uN -r orig/src/libgcrypt.h mod/src/libgcrypt.h
--- a/src/libgcrypt.h	2024-10-16 08:03:11.000000000 +0000
+++ b/src/libgcrypt.h	2026-08-29 21:54:45.336556316 +0000
@@ -61,7 +61,7 @@
 
 #define LIBSSH2_RSA 1
 #define LIBSSH2_RSA_SHA1 1
-#define LIBSSH2_RSA_SHA2 0
+#define LIBSSH2_RSA_SHA2 1
 #define LIBSSH2_DSA 1
 #define LIBSSH2_ECDSA 0
 #define LIBSSH2_ED25519 0
